### PR TITLE
Centralize content type

### PR DESCRIPTION
**Pull request by martingraf4**

### CODE REVIEW

File: app/components/utility/Editables/EditableArray/EditableArray.props.ts [hunk 1]
```diff
@@ -1,10 +1,10 @@
-import { Draft } from 'immer';
+import { Immutable } from 'immer';
 
 import { Editable } from '../../../../packages/library/editor/Editable';
 
 export interface EditableArrayProps<T> {
   array: Editable<T[]>;
   children: (item: Editable<T>, remove: () => void) => JSX.Element;
-  draftNewItem: () => Draft<T>;
+  draftNewItem: () => Immutable<T>;
```
Comment (icefirefly) [inline app/components/utility/Editables/EditableArray/EditableArray.props.ts:8]: Why immutable?

File: app/components/SearchPage/SearchResults.tsx [hunk 2]
```diff
@@ -15,13 +13,10 @@ export default function SearchResults(): JSX.Element {
...
-          <Tile
-            key={uniqueId('searchResults')}
-            content={item}
-            onClick={() => console.log('Test')}
-          />
-        ))}
+        {content.map((item) => {
+          const View = main.contentTypes[item.type].views.ViewSmall;
+          return <View key={item.id} content={item} />;
+        })}
```
Comment (icefirefly) [inline app/components/SearchPage/SearchResults.tsx:19]: Those lines will probably repeat very often. Shorthand form?

File: app/packages/domain/content/ContentEditor/MockContentEditor.ts [hunk 3]
```diff
@@ -35,40 +27,21 @@ export class MockContentEditor
   }
 
   private async contentsWithout(
-    removeContent: ContentID
+    removeContent: ID
```
Comment (icefirefly) [inline app/packages/domain/content/ContentEditor/MockContentEditor.ts:30]: no function of js or immer/lodash for this functionality?

File: app/packages/domain/content/ContentEditor/MockContentEditor.ts [hunk 4]
```diff
@@ -35,40 +27,21 @@ export class MockContentEditor
...
   ): Promise<Immutable<(Content | UnresolvedContent)[]>> {
     return Promise.resolve(this.contents).then((contents) =>
       contents.filter(({ id }) => id !== removeContent)
     );
   }
+
   private newOrEditContentToUnresolvedContent(
-    content: Immutable<NewContent | EditContent>
+    content: Immutable<NewContent | UnresolvedContent>
```
Comment (icefirefly) [inline app/packages/domain/content/ContentEditor/MockContentEditor.ts:38]: long function name, but no other option tbh

File: app/packages/domain/content/ContentTypes/ContentType.ts [hunk 5]
```diff
@@ -1,10 +1,61 @@
...
+    Content
+  >;
+  readonly views: ContentTypeViews<NewContent, UnresolvedContent, Content>;
+}
+
+export interface ContentTypeMetaInfo<ID extends ContentTypeID> {
+  readonly contentTypeId: ID;
+  // TODO make this work with internationalization
+  readonly name: string;
+  readonly pronoun: string;
```
Comment (icefirefly) [inline app/packages/domain/content/ContentTypes/ContentType.ts:26]: What is the pronoun of a content type? The article?

File: app/packages/domain/content/ContentTypes/ContentType.ts [hunk 6]
```diff
@@ -1,10 +1,61 @@
...
+  Content extends UnresolvedContent
+> {
+  Edit(props: EditContentProps<NewContent>): JSX.Element;
+  ViewLarge(props: ViewLargeContentProps<Content>): JSX.Element;
+  ViewSmall(props: ViewSmallContentProps<Content>): JSX.Element;
+}
+
+export interface ViewLargeContentProps<Content> {
+  content: Immutable<Content>;
+}
```
Comment (icefirefly) [inline app/packages/domain/content/ContentTypes/ContentType.ts:53]: Purpose of large and small views?

File: app/packages/domain/content/ContentTypes/FILE/index.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+      filters: new FilterableContentComponentType(main.categorizationProvider),
+    },
+    meta: {
+      contentTypeId: ContentTypeID.FILE,
+      name: 'file',
+      pronoun: 'a',
+      icon: DocumentTextIcon,
+    },
+    views: {
+      Edit: EditFileContent,
```
Comment (icefirefly) [inline app/packages/domain/content/ContentTypes/FILE/index.ts:34]: attributes start with capital letter?

File: app/packages/domain/content/ContentTypes/FILE/types.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import { ContentTypeID } from '../ContentTypeID';
+import {
+  ContentOfAggregateContentTypeComponents,
+  NewContentOfAggregateContentTypeComponents,
+  UnresolvedContentOfAggregateContentTypeComponents,
```
Comment (icefirefly) [inline app/packages/domain/content/ContentTypes/FILE/types.ts:5]: huge names but makes sense

File: app/packages/domain/content/ContentTypes/components/filterable/functions.ts [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+  constructor(
+    private readonly categorizationProvider: CategorizationProvider
+  ) {}
+
+  resolve(
+    unresolved: Immutable<UnresolvedFilterableContentComponent>
+  ): Immutable<FilterableContentComponent> {
+    return {
+      filters: unresolved.filters.map(({ id }) => {
+        const filter = this.categorizationProvider.filters().get(id);
```
Comment (icefirefly) [inline app/packages/domain/content/ContentTypes/components/filterable/functions.ts:28]: shouldn't it be filters?

File: app/packages/domain/content/ContentTypes/helpers/aggregateContentTypes/types/AggregateContentType.ts [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+>;
+
+export type NewContentOfAggregateContentTypeComponents<
+  Components extends AggregateContentTypeComponents,
+  Type extends ContentTypeID
+> = MapComponentsToReturnTypeOfFunction<Components, 'createNew'> & {
+  type: Type;
+};
+
+export type UnresolvedContentOfAggregateContentTypeComponents<
```
Comment (icefirefly) [inline app/packages/domain/content/ContentTypes/helpers/aggregateContentTypes/types/AggregateContentType.ts:31]: huge names but no other option

File: app/packages/domain/main/MockDomainMain.ts [hunk 11]
```diff
@@ -45,10 +47,11 @@ export class MockDomainMain implements EditableDomainMain {
     [this.categorizationEditor, this.categorizationProvider] =
       this.createCategorization();
     this.searchOptionsProvider = this.createSearchOptions();
-    [this.contentEditor, this.contentProvider] = this.createContent();
     [this.externalResourceGroupsEditor, this.externalResourceGroupsProvider] =
       this.createExternalResourceGroups();
     [this.layoutEditor, this.layoutProvider] = this.createLayout();
+    [this.contentEditor, this.contentProvider, this.contentTypes] =
```
Comment (icefirefly) [inline app/packages/domain/main/MockDomainMain.ts:51]: explain this function type

File: app/packages/library/types/assertType.ts [hunk 12]
```diff
@@ -0,0 +1,5 @@
+export function assertType<Base>(): <Extends extends Base>(
```
Comment (icefirefly) [inline app/packages/library/types/assertType.ts:1]: what is the purpose of this?

**Review by icefirefly — COMMENTED** [review]
Lots of test renamings, looks good but very long names sometimes

File: app/components/utility/Editables/EditableArray/EditableArray.props.ts (hunk 1 above)
Comment (martingraf4) [reply app/components/utility/Editables/EditableArray/EditableArray.props.ts:8]: will not be changed directly but we use makeEditable to create new versions of it

**Pull request merged**